src/services/communication/spi.py

```diff
@@ -0,0 +1,146 @@
+import socket, queue
+from threading import Thread
+import abc
+from pin.pinout import Pinout
+from shared_memory import SharedMemory
+
+
+class SPIPeripheral(abc.ABC):
+    '''
+    This is an abstract class that represent an SPI peripheral.
+    The protocol simulation has been implemented using a very simple command protocol:
+    if you are a master, you must send commands to the slave before send any packet to
+    select the slave. The commands are this:
+    - SLAVE_SELECT: this command is sended under the number '1'
+    - SLAVE_DESELECT: this command is sended under the number '2'
+    - NORMAL_PACKET: this command is sended under the number '0',
+        followed by the message intented to send.
+    To the other side, if you are a slave, you must check the command before process the data.
+    Communication itself has been implemented using UDP protocol.
+    Messages are handled using queues and threads, one for transmission and other for reception.
+    '''
+    @abc.abstractmethod
+    def _send(self): pass
+    
+    @abc.abstractmethod
+    def _recv(self): pass
+
+    def __init__(self, ip, port): 
+        self.socket = socket.socket(type=socket.SOCK_DGRAM) # Socket used to send and receive SPI data
+        self.socket.bind((ip, port))
+
+        self.transmission_queue = queue.Queue() # Queue used to store messages that should be sended
+        self.reception_queue = queue.Queue() # Queue used to store received messages
+        self.transmission_thread = Thread(target=self._send , args=("Transmitter")) # Thread that manages data transmission
+        self.reception_thread = Thread(target=self._recv , args=("Receiver")) # Thread that manages data reception
+
+        self.transmission_thread.start()
+        self.reception_thread.start()
+
+    def receive(self) -> bytes: 
+        return self.reception_queue.get
+    
+    @abc.abstractmethod
+    def transmit(self): pass
+
+    def __del__(self):
+        self.socket.close()
+        self.stop = True
+        self.transmission_thread.join()
+        self.reception_thread.join()
+        
+
+class SPIMaster(SPIPeripheral): 
+    def _send(self):
+        while(not self.stop):
+            msg = self.transmission_queue.get
+            total_sent = 0
+            while total_sent < len(msg):
+                sent = self.socket.sendto(msg[sent:], self.send_address)
+                if sent == 0:
+                    raise RuntimeError("Socket connection broken")
+                total_sent += sent
```

Comment:
this and the other methods will crash on socket.sendto and socket.recv no matter the stop flag, as the thread will wait on sendto or recv until the file descriptor is closed and the it will return an error because it cannot read

Reply:
Fixed. Now the socket is closed AFTER the threads join, so it wouldn't block when SPI object is destroyed